### PR TITLE
Refactor ServerMonitor, make 'ert api', 'ert vis' behave

### PR DESCRIPTION
**Pull request by dotfloat**
After a non-insignificant amount of pondering, I've come up with the following API:

First, the old `ServerMonitor` was supposed to be reusable for other servers that we would make, but ended up just being tailored to ERT Storage starting, where code that should be related to `ServerMonitor` living outside of it (eg: ERT 3's service code that starts ERT Storage by polling it every second. No bueno.)

Instead, I have separated `ServerMonitor` into reusable parts. `BaseService` is the part that starts and maintains a service, and `Storage` and `WebvizErt` being subclasses. The basic API of `BaseService` is as follows:

```py
class Service(BaseService):
    ...

with Service.start_server() as service:
    # Server starts asyncroniously, allowing other parts to initalise while the service is starting.
    # `service` is of type Service

    # Explicitly wait for the service to start
    service.wait_until_ready()
    # Implicitly wait for the service to start
    service.fetch_conn_info()  # Returns JSON given to us by the service
    # Connect to the service
    client = Service.connect()
    assert service.fetch_conn_info() == client.fetch_conn_info()

# If the service is started in a different process, or some other part of the code:
client = Service.connect(timeout=10)  # blocks for 10s or until a connection is made
use(client.fetch_conn_info())
```

The interface for ERT Storage is also improved. Instead of having to deal with tokens and baseurls, this handles all of it for you:

```py
with Service.start_server():
    with Service.session() as session:
        # .session() calls wait_until_ready internally.
        # session is a requests.Session type
        session.get("/experiments")  # => requests.Response

    async with Service.async_client() as session:
        # .async_client() calls wait_until_ready internally
        # session is a httpx.AsyncClient type
        await session.get("/experiments")
```

**Issue**
Resolves https://github.com/equinor/ert-storage/issues/154

### CODE REVIEW

File: ert_shared/services/_base_service.py [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+    def __init__(
+        self,
+        service_name: str,
+        exec_args: Sequence[str],
+        timeout: int,
+        set_conn_info: Callable[[ConnInfo], None],
+    ):
+        super().__init__()
+
+        # self._shutdown: bool = False
```
Comment (oysteoh) [inline ert_shared/services/_base_service.py:55]: remove all unused  in this module?

File: ert_shared/services/_webviz_ert_main.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
+import subprocess
```
Comment (oysteoh) [inline ert_shared/services/_webviz_ert_main.py:1]: not used

File: ert_shared/services/_webviz_ert_main.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+def run_webviz_ert():
+    signal.signal(signal.SIGINT, handle_exit)
+    # The entry point of webviz is to call it from command line, and so do we.
+    webviz = shutil.which("webviz")
+    if webviz:
+        with tempfile.NamedTemporaryFile() as temp_config:
+            project_identifier = os.getenv("ERT_PROJECT_IDENTIFIER", os.getcwd())
+            if project_identifier is None:
+                logger.error("Unable to find ERT project!")
```
Comment (oysteoh) [inline ert_shared/services/_webviz_ert_main.py:48]: should this just end in logger error?

Comment (dotfloat) [reply ert_shared/services/_webviz_ert_main.py:48]: Most of this file has been copied directly https://github.com/equinor/webviz-ert/blob/master/webviz_ert/ert_hooks.py

**Comment by oysteoh** [comment]
I have read through and tried to run in various ways - seems to work as I would expect. With regards to documentation, i found the code more or less clean and / or self explanatory with good naming and documentation where needed. But - i don't fully understand everything going on without like really read up on all the functionality. Would have been nice with another pair of eyes also skimming through 👀  in such a large and complex PR @DanSava @mortalisk @frode-aarstad ?

File: ert_shared/plugins/hook_specifications/visualization.py [hunk 4]
```diff
@@ -1,14 +0,0 @@
-from ert_shared.plugins.plugin_manager import hook_specification
-
-
-@hook_specification
-def register_visualization_plugin(handler):
```
Comment (DanSava) [inline ert_shared/plugins/hook_specifications/visualization.py:5]: If you removed this I think you need to remove also 
```
.. autofunction:: ert_shared.plugins.hook_specifications.visualization.register_visualization_plugin

.. autoclass:: ert_shared.plugins.VisualizationPluginHandler
    :members: add_plugin
    :undoc-members:
```
from `docs/rst/manual/experimental/plugin_system.rst`

File: ert_shared/main.py [hunk 5]
```diff
@@ -27,12 +27,28 @@
...
+def run_ert_storage(args):
+    from ert_shared.services.storage import Storage
+
+    with Storage.start_server(res_config=args.config) as server:
+        server.wait()
+
+
+def run_webviz_ert(args):
+    from ert_shared.services.storage import Storage
+    from ert_shared.services.webviz_ert import WebvizErt
```
[the remark this answers is not in the document]
Comment (dotfloat) [reply ert_shared/main.py:44]: Lazy, on-demand import. I could make it go on top but I don't want to load webviz-ert unless it's going to be used.

**Review by oysteoh — APPROVED** [review]
Seems to work as intended. There is still improvements which should be made but this is "good" enough for us to start using and testing it. Awesome work by the way, i think this is a cleaner approach handling the servers.

**Review by oysteoh — CHANGES_REQUESTED** [review]
After cleaning up my env and installing again, i now having trouble running `ert vis` and is getting this error. Not entirely sure why i'm getting it, everything is installed as it should i guess.... 🤔 

```
Traceback (most recent call last):
  File "/Users/OOH/code/github_equinor/ert/ert_shared/services/_webviz_ert_main.py", line 11, in <module>
    from webviz_ert.assets import WEBVIZ_CONFIG
ModuleNotFoundError: No module named 'webviz_ert.assets'; 'webviz_ert' is not a package
```

**Review by oysteoh — APPROVED** [review]
🚀

**Pull request merged**